### PR TITLE
add verbose info for incremental build check

### PR DESCRIPTION
**Pull request by ansyral**
@herohua

### CODE REVIEW

File: src/Microsoft.DocAsCode.Build.Engine/SingleDocumentBuilder.cs [hunk 1]
```diff
@@ -196,14 +196,32 @@ private bool GetCanIncremental(string configHash, string versionName)
                 return false;
             }
             var version = LastBuildInfo.Versions.SingleOrDefault(v => v.VersionName == versionName);
-            if (version == null || configHash != version.ConfigHash)
+            if (version == null)
             {
+                Logger.LogVerbose($"cannot build incrementally because last build didn't contain version {versionName}.");
```
Comment (herohua) [inline src/Microsoft.DocAsCode.Build.Engine/SingleDocumentBuilder.cs]: `cannot` -> `Cannot`?

File: src/Microsoft.DocAsCode.Build.Engine/SingleDocumentBuilder.cs [hunk 2]
```diff
@@ -196,14 +196,32 @@ private bool GetCanIncremental(string configHash, string versionName)
...
                 return false;
             }
-
-            return CurrentBuildInfo.DocfxVersion == LastBuildInfo.DocfxVersion &&
-                CurrentBuildInfo.PluginHash == LastBuildInfo.PluginHash &&
-                CurrentBuildInfo.TemplateHash == LastBuildInfo.TemplateHash;
+            if (configHash != version.ConfigHash)
+            {
+                Logger.LogVerbose($"cannot build incrementally because config changed.");
```
Comment (herohua) [inline src/Microsoft.DocAsCode.Build.Engine/SingleDocumentBuilder.cs]: `$` is not necessary

File: src/Microsoft.DocAsCode.Build.Engine/SingleDocumentBuilder.cs [hunk 3]
```diff
@@ -913,9 +933,14 @@ private bool CanProcessorIncremental(IDocumentProcessor processor, string versio
             {
                 return false;
             }
-            if (!(processor is ISupportIncrementalDocumentProcessor) ||
-                !processor.BuildSteps.All(step => step is ISupportIncrementalBuildStep))
+            if (!(processor is ISupportIncrementalDocumentProcessor))
             {
+                Logger.LogVerbose($"processor {processor.Name} cannot suppport incremental build because the processor doesn't implement ISupportIncrementalDocumentProcessor interface.");
```
Comment (herohua) [inline src/Microsoft.DocAsCode.Build.Engine/SingleDocumentBuilder.cs]: >ISupportIncrementalDocumentProcessor [](start = 137, length = 36)

use nameof

File: src/Microsoft.DocAsCode.Build.Engine/SingleDocumentBuilder.cs [hunk 4]
```diff
@@ -927,13 +952,20 @@ private bool CanProcessorIncremental(IDocumentProcessor processor, string versio
...
                 return false;
             }
             if (cpi.IncrementalContextHash != lpi.IncrementalContextHash)
             {
+                Logger.LogVerbose($"processor {processor.Name} cannot support incremental build because incremental context hash changed.");
+                return false;
+            }
+            if (!new HashSet<ProcessorStepInfo>(cpi.Steps).SetEquals(lpi.Steps))
+            {
+                Logger.LogVerbose($"processor {processor.Name} cannot support incremental build because Steps changed.");
```
Comment (herohua) [inline src/Microsoft.DocAsCode.Build.Engine/SingleDocumentBuilder.cs]: >Steps [](start = 104, length = 5)

`Steps` -> `steps`

File: src/Microsoft.DocAsCode.Build.Engine/SingleDocumentBuilder.cs [hunk 5]
```diff
@@ -196,14 +196,32 @@ private bool GetCanIncremental(string configHash, string versionName)
...
-                CurrentBuildInfo.PluginHash == LastBuildInfo.PluginHash &&
-                CurrentBuildInfo.TemplateHash == LastBuildInfo.TemplateHash;
+            if (configHash != version.ConfigHash)
+            {
+                Logger.LogVerbose($"cannot build incrementally because config changed.");
+                return false;
+            }
+            if (CurrentBuildInfo.DocfxVersion != LastBuildInfo.DocfxVersion)
+            {
+                Logger.LogVerbose($"cannot build incrementally because docfx version changed.");
```
Comment (hellosnow) [inline src/Microsoft.DocAsCode.Build.Engine/SingleDocumentBuilder.cs]: >changed [](start = 85, length = 7)

need we log docfx version before and after?

Comment (ansyral) [inline src/Microsoft.DocAsCode.Build.Engine/SingleDocumentBuilder.cs]: it won't cost much to add the info, i'll add it

---
In reply to: [76950093](https://github.com/dotnet/docfx/pull/655#discussion_r76950093) [](ancestors = 76950093)

**Comment by hellosnow** [comment]
:shipit:

**Comment by herohua** [comment]
:+1:

**Comment by superyyrrzz** [comment]
:shipit:

**Pull request merged**